foxpass-radius-agent.py

```diff
@@ -121,9 +143,76 @@ def two_factor(username):
     if response and response['result'] == 'allow':
         return True
 
-    logger.info("Duo failed")
+    logger.info("Duo mfa failed")
+    return False
+
+def okta_mfa(username):
+    # if Duo is not configured, return success
+    if not get_config_item('okta_hostname') or \
+       not get_config_item('okta_apikey'):
+        logger.info("Okta MFA not configured")
+        return True
+
+    hostname = get_config_item('okta_hostname')
+    api_key = get_config_item('okta_apikey')
+
+    # optional timeout on requests, not related to overall transaction
+    timeout = get_config_item('okta_request_timeout')
+    if not timeout:
+        timeout = 60
+
+    headers = {'Content-type': 'application/json', 'Accept': 'application/json', 'Authorization': 'SSWS %s' % api_key}
+
+    # get the user id from okta
+    url = "https://%s/api/v1/users/%s" % (hostname, username,)
+    resp_json = okta_request(url, headers, timeout)
+
+    if not 'id' in resp_json:
+        logger.info("No Okta user found")
+        return False
+
+    okta_id = resp_json['id']
+
+    # get the factors from okta, look for the push factor
+    url = "https://%s/api/v1/users/%s/factors" % (hostname, okta_id,)
+    resp_json = okta_request(url, headers, timeout)
+    fid = None
+    for factor in resp_json:
+        if factor['provider'] == 'OKTA' and factor['factorType'] == 'push':
+            url = factor['_links']['verify']['href']
+            fid = factor['id']
+            break
+
+    if not fid:
+        logger.info("No Okta push mfa set up")
+        return False
+
+    # start a new verify transaction
+    resp_json = okta_request(url, headers, timeout, post=True)
+
+    # poll for transaction completion
+    while True:
+        if resp_json['factorResult'] == 'SUCCESS':
+            return True
+        elif resp_json['factorResult'] == 'WAITING':
+            url = resp_json['_links']['poll']['href']
+            # sleep for 1 second to rate limit requests
+            time.sleep(1)
```

Comment:
Please read http://www.gevent.org/gevent.monkey.html and make sure time.sleep() is covered by monkey.patch_all() above.

Reply:
it looks like it is covered but i'm not really sure what it does. maybe explain face to face